src/lib/utils/spatial-index.ts

```
@@ -0,0 +1,266 @@
+        }
+      );
+    }
+
+    // Batch add items to grid cells
+    const cellMap = new Map<string, T[]>();
+    items.forEach((item) => {
+      const key = this.getCellKey(item.coordinates);
+      if (!cellMap.has(key)) {
+        cellMap.set(key, []);
+      }
+      cellMap.get(key)!.push(item);
```

Biome (1.9.4)

[error] Forbidden non-null assertion.

Unsafe fix: Replace with optional chain operator ?. This operator includes runtime checks, so it is safer than the compile-only non-null assertion operator


(lint/style/noNonNullAssertion)